File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-2.cpp
class Solution {
private:
    int ROWS;
    int COLS;
    vector<vector<int>> dp;

    int dfs(const vector<vector<int>>& matrix, int r, int c) {
        if (r < 0 || r >= ROWS || c < 0 || c >= COLS) {
            return 0;
        }

        if (dp[r][c] != -1) {
            return dp[r][c];
        }

        int longestPathHere = 1;
        int dirs[] = {0, 1, 0, -1, 0};

        for (int i = 1; i < 5; i++) {
            int nr = r + dirs[i - 1];
            int nc = c + dirs[i];

            if (nr < 0 || nr >= ROWS || nc < 0 || nc >= COLS)
                continue;

            if (matrix[nr][nc] > matrix[r][c])
                longestPathHere = max(longestPathHere, 1 + dfs(matrix, nr, nc));
        }

        return dp[r][c] = longestPathHere;
    }
    
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        ROWS = matrix.size();
        COLS = matrix[0].size();
        dp.assign(ROWS, vector<int>(COLS, -1));

        int longestPathHere, longestPathOverall = 1;
        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                longestPathHere = dfs(matrix, r, c);
                longestPathOverall = max(longestPathHere, longestPathOverall);
            }
        }

        return longestPathOverall;
    }
};

File: Data Structures & Algorithms/maximal-square/submission-2.java
class Solution {
    public int maximalSquare(char[][] matrix) {
        int ROWS = matrix.length, COLS = matrix[0].length, ans = 0;

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if ((r == 0 || c == 0) && matrix[r][c] == '1') {
                    ans = Math.max(ans, 1);
                }
                else if (matrix[r][c] == '1') {
                    matrix[r][c] = (char)('1' + Math.min((int)(matrix[r - 1][c - 1] - '0'), 
                        Math.min((int)(matrix[r - 1][c] - '0'), (int)(matrix[r][c - 1]) - '0')));
                    ans = Math.max(ans, (int)(matrix[r][c] - '0'));
                }
            }
        }

        return ans * ans;
    }
}